### PR TITLE
allow interactive tools on normal workers for oubb training

### PR DESCRIPTION
**Pull request by mira-miracoli**

### CODE REVIEW

**Review by sanjaysrikakulam — CHANGES_REQUESTED** [review]
The following should work, as we only want to overwrite the default [submit_requirements](https://github.com/usegalaxy-eu/infrastructure-playbook/blob/master/files/galaxy/tpv/interactive_tools.yml#L9) for the ITs for the OUBB training session. Instead of the `GalaxyGroup == "interactive"` we want it to say `GalaxyGroup == "training-oubb"`

```
training_roles = [r.name for r in user.all_roles() if not r.deleted and "training" in r.name]
oubb_roles = [r for r in training_roles if "oubb" in r]

if oubb_roles:
    # If there is an 'oubb' role, set the submit requirements accordingly
    training_expr = 'GalaxyGroup == "%s"' % oubb_roles[0]
    entity.params['submit_requirements'] = 'GalaxyDockerHack == True && %s' % training_expr
```

The goal is for the users in the `training_oubb` role to direct their ITs to the training dedicated infrastructure instead of the interactive workers.

I briefly tested it:

```python
class MockRole:
    def __init__(self, name, deleted):
        self.name = name
        self.deleted = deleted

class MockUser:
    def __init__(self, roles):
        self.roles = roles

    def all_roles(self):
        return self.roles

user = MockUser(roles=[
    MockRole(name="training-oubb", deleted=False),
    MockRole(name="training-general", deleted=False)
])

training_roles = [r.name for r in user.all_roles() if not r.deleted and "training" in r.name]
oubb_roles = [r for r in training_roles if "oubb" in r]

params = {'submit_requirements':'GalaxyDockerHack == True && GalaxyGroup == "interactive"'}    # This is the same as the one in line 9 in the interactive_tools.yml file.
if oubb_roles:
    # If there is an 'oubb' role, set the submit requirements accordingly
    training_expr = 'GalaxyGroup == "%s"' % oubb_roles[0]
    params['submit_requirements'] = 'GalaxyDockerHack == True && %s' % training_expr

print(params)
```

I can test the rules in ESGs TPV if you would like.

**Comment by sanjaysrikakulam** [comment]
This is a temp rule and valid between 14/10/2024 and 21/10/2024. The `submit_requirements` should only be modified for the users in the role `training-oubb`.

**Comment by mira-miracoli** [comment]
Yes I I know, but what I mean with
> Maybe we should add the '(GalaxyGroup == "compute") || (%s)' % training_expr [part](https://github.com/usegalaxy-eu/infrastructure-playbook/blob/305c60ef887dd8bd0828fc4564f3dda340a3d63d/files/galaxy/tpv/tool_defaults.yml#L29C1-L29C146), because from what I see now the jobs can really only go to the training vms and not to normal queue.

is that when their VM is full, they also can not schedule interactive jobs on the normal interactive nodes.
So this part could be
~~~
'GalaxyDockerHack == True && %s || GalaxyGroup == "interactive"' % training_expr 
~~~

**Comment by sanjaysrikakulam** [comment]
> Yes I I know, but what I mean with
> 
> > Maybe we should add the '(GalaxyGroup == "compute") || (%s)' % training_expr [part](https://github.com/usegalaxy-eu/infrastructure-playbook/blob/305c60ef887dd8bd0828fc4564f3dda340a3d63d/files/galaxy/tpv/tool_defaults.yml#L29C1-L29C146), because from what I see now the jobs can really only go to the training vms and not to normal queue.
> 
> is that when their VM is full, they also can not schedule interactive jobs on the normal interactive nodes. So this part could be
> 
> ```
> 'GalaxyDockerHack == True && %s || GalaxyGroup == "interactive"' % training_expr 
> ```

I compared the allocated resources and checked the number of participants, thinking it might be optional. However, it's a good catch; it adds more flexibility, which is nice. Please feel free to update the `training_expr` and push the changes.

File: files/galaxy/tpv/interactive_tools.yml [hunk 1]
```diff
@@ -15,7 +14,14 @@ tools:
...
           Interactive tools require registration. Please log-in or register on https://usegalaxy.eu/login
-
+      - if: user is not None
+        execute: |
+          training_roles = [r.name for r in user.all_roles() if not r.deleted and "training" in r.name]
+          oubb_roles = [r for r in training_roles if "oubb" in r]
+          if oubb_roles:
+              # If there is an 'oubb' role, set the submit requirements accordingly
+              training_expr = 'GalaxyGroup == "%s"' % oubb_roles[0]
+              params['submit_requirements'] = 'GalaxyDockerHack == True && GalaxyGroup == "interactive" || %s' % training_expr
```
Comment (sanjaysrikakulam) [inline files/galaxy/tpv/interactive_tools.yml:24]:
```suggestion
              entity.params['submit_requirements'] = 'GalaxyDockerHack == True && GalaxyGroup == "interactive" || %s' % training_expr
```

File: files/galaxy/tpv/interactive_tools.yml [hunk 2]
```diff
@@ -6,7 +6,6 @@ tools:
     params:
       docker_volumes: $defaults
       container_monitor_result: callback
-      submit_requirements: 'GalaxyDockerHack == True && GalaxyGroup == "interactive"'
```
Comment (sanjaysrikakulam) [inline files/galaxy/tpv/interactive_tools.yml:9]: This should exist for all other users not part of the `oubb` role, or am I missing something?

Comment (mira-miracoli) [reply files/galaxy/tpv/interactive_tools.yml:9]: sorry I copied the code from your snipped from the wrong part

**Comment by mira-miracoli** [comment]
Linting fails because there is no test user, I guess

**Review by sanjaysrikakulam — APPROVED** [review]
Looks good, thank you!

**Pull request merged**